### ISSUE TITLE
Getting java.lang.NoClassDefFoundError when using `redefine` and `intercept(MethodDelegation.to(SomeClass.class))`

### ISSUE THREAD

**Issue by ashleyfrieze** [raphw/byte-buddy#1517]
Here's the code I'm trying to write:

```java
// inside the static constructor of class EnvironmentVariableMocker

            ByteBuddyAgent.install();
            var byteBuddy = new ByteBuddy();
            byteBuddy.redefine(Class.forName("java.lang.ProcessEnvironment"))
                .method(namedOneOf("getenv", "environment", "toEnvironmentBlock"))
                .intercept(MethodDelegation.to(EnvironmentVariableMocker.class))
                .make()
                .load(
                    EnvironmentVariableMocker.class.getClassLoader(),
                    ClassReloadingStrategy.fromInstalledAgent());
```

I'm attempting to intercept some key methods of `ProcessEnvironment` to write a test utility:  https://github.com/webcompere/system-stubs/blob/threads-example/system-stubs-core/src/main/java/uk/org/webcompere/systemstubs/environment/EnvironmentVariableMocker.java

_Aside: I have previously used Mockito to mock these calls, but Mockito deliberately prevents static mocking affecting multiple threads._

The problem is that ByteBuddy happily accepts the request to use this class as an alternative implementation for the methods in question, and it's clearly making ProcessEnvironment do something different after reloading it.

However, we get this error:

```
java.lang.NoClassDefFoundError: uk/org/webcompere/systemstubs/environment/EnvironmentVariableMocker

	at java.base/java.lang.ProcessEnvironment.getenv(ProcessEnvironment.java)
	at java.base/java.lang.System.getenv(System.java:1155)
```

So the reloaded `ProcessEnvironment` class cannot see the code I want to delegate to.

I tried using an object instance, but it didn't work. And I tried splitting out the behaviour into a sibling class, in case this was something to do with the class not being allowed to nominate itself as the new implementation. This failed with the same error.

Am I doing something wrong here?

**Comment by raphw** [comment]
The class you are instrumenting is on the boot loader. Your agent is on the system loader. The system class is not visible to the boot class.

Either you can inject the dispatcher into the boot class, or you inline the code in question using Advice.

**Comment by ashleyfrieze** [comment]
@raphw thanks so much for getting back to me so quickly. I'm new to ByteBuddy. Do you have any worked examples of either of these I could follow?

**Comment by raphw** [comment]
Have s look st ClassInjector and Advice. Both document themselves in their javadoc.

**Comment by ashleyfrieze** [comment]
@raphw - thanks for your help. We can close this one.

For reference, the final solution I chose was:

- build the class that's used for intercepting in a separate jar
- use it at compile time as a `provided` dependency - i.e. it's expected to be put into the classpath by someone else at runtime
- bring that jar into the project as a resource
- inject the jar with `Instrumentation` into the bootClassLoader using `appendToBootstrapClassLoaderSearch`
- then use `.intercept(MethodDelegation.to(ProcessEnvironmentInterceptor.class))` on ByteBuddy to attach it to `ProcessEnvironment`

I tried other variants. I had a lot of trouble getting an Advice-based class to work with the static methods, owing to errors about not being able to call a super class method. In the end, I think I used a transformer to get it to work, but then hit the same problem that any code in the Advice that relied on classes not already in the boot class loader would get class not found errors.

The "learn" section of ByteBuddy.net doesn't contain too much information about recipes for ClassInjector and Advice, and while the JavaDoc is good for those who already understand the fundamentals, I found it tricky to work out the correct way to use things.

In the end, I reverse engineered the good work you did on Mockito-inline in order to see what might be possible, and I think my mechanism for injecting the jar file is similar to that used in Mockito, though perhaps dumbed down so I didn't need to construct the `.jar` from a `.class` file.
